File: 95 (Unique Binary Search Trees II)/cpp/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> solve(int start, int end){
        vector<TreeNode*> res;
        if (start>end){
            res.push_back(nullptr);
            return res;
        }

        for (int i = start;i<=end;i++){
            auto left_side=  solve(start,i-1);
            auto right_side = solve(i+1,end);

            for(auto left_node: left_side){
                for(auto right_node: right_side){
                    TreeNode* root = new TreeNode(i);
                    root->left = left_node;
                    root->right = right_node;
                    res.push_back(root);
                }
            }
        }
        return res;
    }
    vector<TreeNode*> generateTrees(int n) {
        return solve(1,n);
    }
};

File: 95 (Unique Binary Search Trees II)/python/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        
        def solve(start,end):
            if start>end: return [None]
            res = []
            for i in range(start,end+1):
                left_side = solve(start,i-1)
                right_side = solve(i+1,end)
            
                for left_node in left_side:
                    for right_node in right_side:
                        root = TreeNode(i)
                        root.left = left_node
                        root.right = right_node
                        res.append(root)
            return res
        return solve(1,n)